### PR TITLE
Sec release

### LINKED ISSUE #402: Secondary release areas [com1DFAdev]
Add the option to use secondary release areas

### PR DESCRIPTION
**Pull request by matthiasto**
Enable secondary release area: 
- possible to provide a sec release shp file in SECREL folder (only one allowed features should not overlap)
- read the secondary release feature and initialize it (done at the same time as the primary release (maybe we want to do this only if the release is hit?)
- add particles to the computation as soon as the sec rel area is hit by the flowing avalanche

I also fixed the small bug for exporting time step results that was not working for particles

fix #402

### CODE REVIEW

File: avaframe/out3Plot/outAIMEC.py [hunk 1]
```diff
@@ -150,7 +150,7 @@ def visuRunoutComp(rasterTransfo, resAnalysis, newRasters, cfgSetup, cfgPath, cf
         ax.plot(meanVal[1, :], s, '-b', label='Mean Simulation')
 
         ax.set_title(titleVal + 'distribution along path')
-        ax.legend(loc=4)
+        ax.legend(loc=1)
```
Comment (fso42) [inline avaframe/out3Plot/outAIMEC.py:153]: maybe use loc='best' here? 

> The string 'best' places the legend at the location, among the nine locations defined so far, with the minimum overlap with other drawn artists.

**Review by fso42 — APPROVED** [review]

File: avaframe/com1DFAPy/com1DFA.py [hunk 2]
```diff
@@ -492,17 +533,29 @@ def initializeSimulation(cfg, demOri, releaseLine, entLine, resLine, logName, re
...
+
+    particles['secondaryReleaseInfo'] = secondaryReleaseInfo
     # initialize entrainment and resistance
     # get info of simType and whether or not to initialize resistance and entrainment
     simTypeActual = cfgGen['simTypeActual']
     rhoEnt = cfgGen.getfloat('rhoEnt')
     hEnt = cfgGen.getfloat('hEnt')
+    entLine = inputSimLines['entLine']
     entrMassRaster, reportAreaInfo = initializeMassEnt(demOri, simTypeActual, entLine, relRaster, reportAreaInfo)
+    resLine = inputSimLines['resLine']
```
Comment (awirb) [inline avaframe/com1DFAPy/com1DFA.py:558]: you could also just put inputSimLines['resLine'] as input parameter

File: avaframe/com1DFAPy/com1DFA.py [hunk 3]
```diff
@@ -1264,12 +1320,19 @@ def prepareArea(releaseLine, dem, relThList=''):
         dictionary with dem information
     relThList: list
         release thickness values for all release features
+    combine : Boolean
+        if True sum upe the rasters in the area list to return only 1 raster
```
Comment (awirb) [inline avaframe/com1DFAPy/com1DFA.py:1324]: sum up

File: avaframe/com1DFAPy/com1DFA.py [hunk 4]
```diff
@@ -1482,6 +1547,44 @@ def plotContours(fig, ax, particles, dem, data, Cmap, unit, last=False):
...
+                # create secondary release area particles
+                log.info('Initializing secondary release area feature %s' % secRelRasterName)
+                secRelParticles, fields = initializeParticles(cfg, secRelRaster, dem)
+
+                # # save particles and fields for the initial time step to file
+                # outDirData = os.path.join(outDir, 'secondaryReleaseArea', secRelRasterName)
+                # exportFields(cfg, [0.], [fields], demOri, outDirData, secRelRasterName)
+                # if 'particles' in cfg['resType']:
+                #     # export particles dictionaries of saving time steps
+                #     savePartToPickle([secRelParticles], outDirData)
```
Comment (awirb) [inline avaframe/com1DFAPy/com1DFA.py:1571]: still needed?

File: avaframe/in1Data/getInput.py [hunk 5]
```diff
@@ -202,10 +203,10 @@ def getInputDataCom1DFAPy(avaDir, cfg, flagDev=False):
     if flagDev is True:
         releaseDir = 'devREL'
         relFiles = glob.glob(inputDir+os.sep + releaseDir+os.sep + '*.shp')
-    elif cfg['releaseScenario'] != '':
+    elif cfg['FLAGS']['releaseScenario'] != '':
```
Comment (awirb) [inline avaframe/in1Data/getInput.py:206]: should this be in the flags? it is not a boolean..

**Review by awirb — CHANGES_REQUESTED** [review]

File: avaframe/in1Data/getInput.py (hunk 5 above)
Comment (matthiasto) [reply avaframe/in1Data/getInput.py:206]: no this is if you want to specify a specific release. This was already in the ini flag section, I did not change this

Comment (matthiasto) [reply avaframe/in1Data/getInput.py:206]: It was called like this before:
demFile, relFiles, entFiles, resFile, entResInfo = gI.getInputDataCom1DFAPy(
        avaDir, cfg['FLAGS'], flagDev)
    and now like this:
demFile, inputSimFiles = gI.getInputDataCom1DFAPy(avaDir, cfg, flagDev)

File: avaframe/com1DFAPy/com1DFACfg.ini [hunk 6]
```diff
@@ -78,6 +78,10 @@ subgridMixingFactor = 100.
 
 # release depth
 relTh = 1.
+# Add secondary release area
+secRelArea = False
+# secondary area release depth
+secondaryRelTh = 0.5
```
[the remark this answers is not in the document]
Comment (matthiasto) [reply avaframe/com1DFAPy/com1DFACfg.ini:85]: Why make it more proiminent than the relTh or other parameters? it is set to false as default anyways, so if you want to have it, you look at the doc on how to activate it no?

File: avaframe/com1DFAPy/com1DFA.py (hunk 4 above)
Comment (matthiasto) [reply avaframe/com1DFAPy/com1DFA.py:1571]: Depends on if we want to save the secRelease particles initial state

File: avaframe/com1DFAPy/com1DFACfg.ini (hunk 6 above)
[the remark this answers is not in the document]
Comment (awirb) [reply avaframe/com1DFAPy/com1DFACfg.ini:85]: but I still think the seRelArea corresponds to the flags section at least

File: avaframe/com1DFAPy/com1DFA.py [hunk 7]
```diff
@@ -1787,25 +1796,33 @@ def splitPart(cfg, particles, dem):
...
+    Parameters
+    ----------
+    particles1 : dict
+        first particles dictionary
+    particles2 : dict
+        second particles dictionary
+
+    Returns
+    -------
+    particles1 : dict
```
Comment (fso42) [inline avaframe/com1DFAPy/com1DFA.py:1811]:
```suggestion
    particles : dict
```

**Review by fso42 — CHANGES_REQUESTED** [review]

File: avaframe/in1Data/getInput.py [hunk 8]
```diff
@@ -182,73 +156,110 @@ def getInputDataCom1DFAPy(avaDir, cfg, flagDev=False):
...
-            raise
-        entResInfo['flagEnt'] = 'Yes'
+    entFile, entResInfo['flagEnt'] = getAndCheckInputFiles(inputDir, 'ENT', 'Entrainment')
 
     # Initialise DEM
     demFile = getDEMPath(avaDir)
 
     # return DEM, first item of release, entrainment and resistance areas
-    return demFile, relFiles, entFiles[0], resFiles[0], entResInfo
+    inputSimFiles = {'relFiles': relFiles, 'secondaryReleaseFile': secondaryReleaseFile,
```
Comment (awirb) [inline avaframe/in1Data/getInput.py:216]: suggestion: put dem also as key in inputSimFiles

**Review by awirb — CHANGES_REQUESTED** [review]

File: avaframe/com1DFAPy/com1DFA.py [hunk 9]
```diff
@@ -475,15 +516,20 @@ def initializeSimulation(cfg, demOri, releaseLine, entLine, resLine, logName, re
 
     # -----------------------
     # Initialize mesh
+    log.info('Initializing Mesh')
     demOri, dem = initializeMesh(cfgGen, demOri, methodMeshNormal)
+
     # ------------------------
+    log.info('Initializing main release area')
```
Comment (awirb) [inline avaframe/com1DFAPy/com1DFA.py:523]: maybe debug level - only if second release area is initialised info level?

**Review by awirb — CHANGES_REQUESTED** [review]

File: avaframe/in3Utils/geoTrans.py [hunk 10]
```diff
@@ -655,10 +655,11 @@ def checkOverlap(toCheckRaster, refRaster, name, crop=False):
     if mask.any():
         if crop:
             toCheckRaster[mask] = 0
-            message = name + ' area features are overlaping - removing the overlaping part'
-            log.warning('%s' % message)
+            message = '%s area feature overlaping with %s - removing the overlaping part' % (nameToCheck, nameRef)
```
Comment (fso42) [inline avaframe/in3Utils/geoTrans.py:658]:
```suggestion
            message = '%s area feature overlapping with %s - removing the overlapping part' % (nameToCheck, nameRef)
```

File: avaframe/in3Utils/geoTrans.py [hunk 11]
```diff
@@ -655,10 +655,11 @@ def checkOverlap(toCheckRaster, refRaster, name, crop=False):
...
+            log.warning(message)
         else:
-            message = name + ' area features are overlaping - this is not allowed'
-            log.error('%s' % message)
+            message = '%s area features overlapingwith %s - this is not allowed' % (nameToCheck, nameRef)
```
Comment (fso42) [inline avaframe/in3Utils/geoTrans.py:661]:
```suggestion
            message = '%s area features overlapping with %s - this is not allowed' % (nameToCheck, nameRef)
```

File: avaframe/com1DFAPy/com1DFA.py [hunk 12]
```diff
@@ -475,15 +516,20 @@ def initializeSimulation(cfg, demOri, releaseLine, entLine, resLine, logName, re
 
     # -----------------------
     # Initialize mesh
+    log.info('Initializing Mesh')
```
[the remark this answers is not in the document]
Comment (matthiasto) [reply avaframe/com1DFAPy/com1DFA.py:519]: maybe we do this in a separate PR where we clean all the log prints

**Review by fso42 — APPROVED** [review]

**Pull request merged**